### ISSUE TITLE
Tutorial 6: Multihead Attention

### ISSUE THREAD

**Issue by Hannibal046** [bentrevett/pytorch-seq2seq#148]
Hello,

About tutorial 6: Attention is all you need.
I noticed that in the `MultiHeadAttentionLayer` implementation, to get Q from `[batch size, query len, hid dim]`  to `[batch size, n heads, query len, head dim]`, the code writes:
```python
Q = Q.view(batch_size, -1, self.n_heads, self.head_dim).permute(0, 2, 1, 3)
```
Rather than:
```python
Q = Q.view(batch_size, self.n_heads, -1, self.head_dim)
```
And I did a little experiment to find out that
```python
x = torch.arange(6).view(2,3)
(x.permute(1,0) == x.view(3,2)) == False
```
I just want to know that when `ndim` of a tensor is  larger than 2, things just get abstract. How to write the correct code about the high dimension tensor? I am really curious about the that. 
When reading about the `MultiHeadAttentionHeadLayer`, I feel like that I will never write code as correct as it is.

Really looking forward to your reply!
@bentrevett

**Comment by bentrevett** [comment]
@Hannibal046 Sorry for the late reply, I have been away on Christmas break. I'm not sure of your exact question but I'll just ramble on through how I think about the equations so feel free to reply if anything needs clarifying. 

The initial Q/K/V shape is `[batch size, query len, hid dim]`, we want to split the last dimension, the `hid dim` one, into `n_heads` equal chunks, of size `head_dim`, which we do with `.view(batch_size, -1, self.n_heads, self.head_dim)`. If we did `.view(batch_size, self.n_heads, -1, self.head_dim)` as you suggested this would try and split across the sequence length, which we don't want.

Let's try and verify the `.view` in the code does what we want:

```python
>>> batch_size = 2
>>> seq_len = 3
>>> hid_dim = 4
>>> x = torch.arange(batch_size * seq_len * hid_dim).view(batch_size, seq_len, hid_dim)
>>> x
tensor([[[ 0,  1,  2,  3], #these three are the first batch element, with seq_len rows and hid_dim columns
         [ 4,  5,  6,  7],
         [ 8,  9, 10, 11]],

        [[12, 13, 14, 15], #these three are the second batch element
         [16, 17, 18, 19], #again, we have seq_len rows and hid_dim columns
         [20, 21, 22, 23]]])
```

Now to split the `hid_dim` dimension:

```python
>>> n_heads = 2
>>> head_dim = hid_dim // n_heads
>>> x.view(batch_size, -1, n_heads, head_dim)
tensor([[[[ 0,  1], #first batch element, first sequence element, first head
          [ 2,  3]], #first batch element, first sequence element, second head
 
         [[ 4,  5], #first batch element,second sequence element, first head
          [ 6,  7]], #first batch element, second sequence element, second head

         [[ 8,  9], #first batch element, third sequence element, first head
          [10, 11]]], #first batch element, third sequence element, second head


        [[[12, 13], #second batch element, first sequence element, first head
          [14, 15]], #second batch element, first sequence element, second head

         [[16, 17], #etc etc etc
          [18, 19]],

         [[20, 21],
          [22, 23]]]])
```

Looks good to me!

Now have reshaped our tensor from `[batch size, seq len, hid dim]` to `[batch size, seq len, n heads, head dim]`. Then we permute so we get `[batch size, n heads, seq len, head dim]`. But why do we permute? Well, when we do the matrix multiplication between the Q and K we want to see how much each `head_dim` of Q "matches" all of the `head_dims` in K. Thus we have to do our matrix multiplication across the sequence length. Thus we want the `seq_len` and `head_dim` to be the last two dimensions of each tensor - so we can multiply them together - hence we permute!

We actually need to permute K again to swap the last to dimensions around so we're multiplying a `[batch size, n heads, query len, head dim]` tensor by a `[batch size, n heads, head dim, key len]` tensor because we want a `[batch size, n heads, query len, key len]` output which contains the values of how much each of the `head_dims` matches the other `head_dims` across each of the lengths. 

Later on we multiply this by V, aftering applying a softmax to go from "energy" to "attention", which means we have a matrix multiplication of `[batch size, n heads, query len, key len]` by a `[batch size, n heads, value len, head dim]` to get a `[batch size, n heads, query len, head dim]` output. Note that this only works because in multi-head attention layer you must **always** have the key length equal to the value length. This matrix multiplication is calculating a weighted sum of all the head dims across the value sequence length, where the weights are the values of attention. 

We now have a `[batch size, n heads, query len, head dim]` tensor, where each `head_dim` is a weighted sum of the value tensor, where the weights are calculated by how much the query "matches" the "key" - this is commonly thought of in terms of a "soft" dictionary access. The last thing we need to do is make sure everything goes back to the correct shape, so we permute to get the `n_head` and `head_dim` dimensions next to each other and then `view` to merge them together again.

The general way of figuring this stuff out is by reading other implementations and then stepping through the maths. I do a lot of matrix multiplications on paper just to try and understand things. I also think http://matrixmultiplication.xyz/ is useful sometimes.

Not sure if this is useful to you in any way, but let me know!

**Comment by Hannibal046** [comment]
Wow，Thanks sooooo much for  detailed explanation. I am sorry for my so-late reply , but i did not notice the github notification of your reply,my bad.
Now, I have understood what you mean, it is of so much insight of your tutorial. My all NLP  knowledge starts from your repository， that is amazing . Do you have a plan to release code like beam search and pointer generator? I will stick to it !!
BTW, I am also looking forward to your new project of `easynlp`
